src/projection.jl

```diff
@@ -352,13 +353,31 @@ function (project::ProjectTo{SparseVector})(dx::AbstractArray)
         reshape(dx, project.axes)
     end
     nzval = map(i -> project.element(dy[i]), project.nzind)
-    n = length(project.axes[1])
-    return SparseVector(n, project.nzind, nzval)
+    return SparseVector(length(dx), project.nzind, nzval)
+end
+function (project::ProjectTo{SparseVector})(dx::SparseVector)
+    size(dx) == map(length, project.axes) || throw(_projection_mismatch(project.axes, size(dx)))
+    # When sparsity pattern is unchanged, all the time is in checking this,
+    # perhaps some simple hash/checksum might be good enough?
+    samepattern = project.nzind == dx.nzind
+    # samepattern = length(project.nzind) == length(dx.nzind) 
+    if eltype(dx.nzval) <: project_type(project.element) && samepattern
+        return dx
```

Comment:
I added some faster paths for sparse arrays. There might be a lot of room to do better. If both `x` and `dx` are sparse, but different, should the projection be to `x`'s nonzero values, or to the intersection?
```julia
julia> using SparseArrays

julia> sv = sprand(10^4,0.1); pv = ProjectTo(sv);

julia> @btime ProjectTo($sv);
  0.875 ns (0 allocations: 0 bytes)

julia> @btime $pv($(rand(length(sv))));
  476.403 ns (1 allocation: 7.94 KiB)

julia> @btime $pv($(sprand(length(sv), 0.1)));

julia> @btime $pv($sv);
  14.458 μs (1 allocation: 7.94 KiB)  # with generic
  220.393 ns (1 allocation: 32 bytes) # with if project.nzind == dx.nzind
  5.167 ns (1 allocation: 32 bytes)   # with if length(project.nzind) == length(dx.nzind)

julia> sm = sprand(10^3, 10^3, 0.01); pm = ProjectTo(sm);

julia> @btime $pm($(rand(size(sm)...)));
  10.916 μs (2 allocations: 77.30 KiB)

julia> @btime pm($(sprand(1000,1000, 0.01)));
  73.791 μs (5 allocations: 77.44 KiB)

julia> @btime $pm($sm);
  67.958 μs (3 allocations: 77.34 KiB)  # naiive alg
  2.190 μs (0 allocations: 0 bytes)     # careful check
  3.625 ns (0 allocations: 0 bytes)     # check length & last
```

Reply:
> should the projection be to x's nonzero values, or to the intersection?

both are legal.
`x`s is all that is required, but the differential is allowed to have "more structure" than the target space,
so if the intersection is faster then can do that.

Reply:
> should the projection be to x's nonzero values, or to the intersection?

For `y = ProjectTo(x)(dx)`

The requirement is that if `x` has a structural zero at a given postition, then `y` must also.
But `y` can also have more structural zeros then `x`
So the zeros for `y` can be either the same as the zeros of `x` or the union of the zeros in `x` and `dx`.

So now to flip that logic around to talk about nonzeros:
the nonzeros in `y` must at most be all the nonzeros in `x`.
But if can be fewer.
It could instead be the intersection of the nonzeros of `x` and `dx`

Reply:
Yes, this is my thinking too. Although there's a chance that whatever consumes `y` would prefer exactly the same nonzero indices, over a slightly different smaller set which it has to think about.

`intersect(::Vector{Int}, ::Vector{Int})` is also not so quick. It's possible that this should store some BitSet or something.